spec/message_spec.rb

```diff
@@ -0,0 +1,67 @@
+RSpec.describe "message encoding" do
+  # message Sally {
+  #   required string mood = 1;
+  #   optional uint32 age  = 2;
+  # }
+  class Sally
+    def initialize(attributes)
+      @attributes = {}
+      attributes.each do |attribute, value|
+        public_send("#{attribute}=", value)
+      end
+    end
+
+    ## Attributes
+
+    def age
+      @attributes[:age]
+    end
+
+    def age=(age)
+      raise ArgumentError, "age is not a number" unless age.respond_to?(:to_i)
+      raise ArgumentError, "age is too large" unless age.to_i <= 4_294_967_295
```

Comment:
this is the point where I am not sure if we want to do the coercion in the setter (and if we do we should only coerce 1 time and then check the bounds)

I do like the range bounds being generated and explicit vs abstracted away into constants as they won't change for the type